docs/extensibility/custom-resources.md

```diff
@@ -0,0 +1,432 @@
+
+```csharp
+app.MapGet("/subscribe", async (SmtpClient smtpClient, string email) => {
+    var message = new MailMessage("newsletter@yourcompany.com", email);
+    message.Subject = "Welcome to our newsletter!";
+    message.Body = "Thank you for subscribing to our newsletter!";
+    await smtpClient.SendMailAsync(message);
+});
+
+app.MapGet("/unsubscribe", async (SmtpClient smtpClient, string email) => {
+    var message = new MailMessage("newsletter@yourcompany.com", email);
+    message.Subject = "You are unsubscribed from our newsletter!";
+    message.Body = "Sorry to see you go. We hope you will come back soon!";
+    await smtpClient.SendMailAsync(message);
+});
+```
+
+Once the `Program.cs` file is updated launch the AppHost and use your browser, or `curl` to hit the following URLs:
+
+```bash
+# Remember to replace the port to match your local environment.
+curl https://localhost:7289/subscribe?email=test@test.com
+curl https://localhost:7289/unsubscribe?email=test@test.com
+```
+
+If those API calls return a successful response then you should be able to click on the `maildev` resource the dashboard and the MailDev UI will show the emails that have been sent to the SMTP endpoint.
+
+![E-mails visible in the MailDev UI](./media/maildev-emails.png)
+
+## Technical details
+
+### IValueProvider and IManifestExpressionProvider
+
+In the code above the `MailDevResource` had two properties. `SmtpEndpoint` and `ConnectionStringExpression`. The types of these properties were <xref:Aspire.Hosting.ApplicationModel.EndpointReference> and <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> respectively. These types are among several which are used throughout .NET Aspire to represent configuration data which is not finalized until the .NET Aspire application is either run or published to the cloud via a tool such as `azd` (Azure Developer CLI).
+
+The fundamental problem that these types help solve is deferring resolution of concrete configuration information until _all_ of the information is available.
+
+For example, in the `MailDevResource` example we expose a property called `ConnectionStringExpression` as required by the <xref:Aspire.Hosting.ApplicationModel.IResourceWithConnectionString> interface. The type of the property is <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression> and is created by passing in an interpolated string to the <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression.Create> method.
+
+```csharp
+public ReferenceExpression ConnectionStringExpression => ReferenceExpression.Create(
+        $"smtp://{SmtpEndpoint.Property(EndpointProperty.Host)}:{SmtpEndpoint.Property(EndpointProperty.Port)}"
+        );
+```
+
+The signature for the <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression.Create> method is as follows:
```

Comment:
@mitchdenny - you asked why these were wrong, when you see the published API docs that correspond to this xref, you can verify the name matches the [HTML's](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.applicationmodel.referenceexpression.create?view=dotnet-aspire-8.0) `content` attribute value of the <head><meta name="ms.assetid"></meta></head>`. In this case, you'd see this:

```html
<meta name="ms.assetid" content="Aspire.Hosting.ApplicationModel.ReferenceExpression.Create*">
```

And that `*` was what you're missing, `%2A`.

```suggestion
The signature for the <xref:Aspire.Hosting.ApplicationModel.ReferenceExpression.Create%2A> method is as follows:
```